### DISCUSSION TITLE
Input parameters not taking nullability into consideration

### DISCUSSION THREAD

**Discussion by justinrassier** [giacomocavalieri/squirrel#23] [General]
I have an update statement that sets a nullable text field:
```sql
UPDATE my_table
SET my_nullable_field = $1
WHERE id = $2
```
But the generated function from squirrel doesn't take into consideration the nullable field and requires me to pass in a `String` instead of an `Option(String)`
```gleam
pub fn update_exam_instance_question(db, arg_1, arg_2, arg_3) {
  let decoder = decode.map(decode.dynamic, fn(_) { Nil })

  "UPDATE my_table
SET my_nullable_field = $1 
WHERE id = $2
"
  |> pgo.execute(
    db,
    [pgo.text(arg_1) // <--   expected  pgo.nullable(pgo.text, arg_1)
   , pgo.int(arg_2))]
    decode.from(decoder, _),
  )
}
```

I'm not sure if there is any way to hint to squirrel that this should be nullable or not

**Comment by giacomocavalieri** [comment]
Ah that's interesting, I need to look a bit into it! There probably is a way to tell if a field in an insert/update query is a nullable column 🤔

**Comment by justinrassier** [comment]
Doing a little searching, it seems like the query itself can't tell you anything about the nullability, but you can query the schema and get that info
```sql
SELECT column_name, is_nullable
FROM information_schema.columns
WHERE table_name = 'my_table'
  AND column_name = 'my_nullable_field';
```
```md
 column_name      | is_nullable
------------------|-------------
 my_nullable_field| YES

```

**Comment by giacomocavalieri** [comment]
Unfortunately I cannot really use that to implement this because `squirrel` doesn't implement a sql parser. I can only use the info the Postgres server sends me once I ask it to prepare the query and it looks like it's not carrying around any info about the parameter's origin (if it, like in this example, a nullable column): I only get its base non-nullable type, `text` in your example.

So I'm not sure this can actually be implemented, I'll have to look more into how sqlx does it (if it actually does it at all)

**Comment by hfjallemark** [comment]
I believe sqlx runs an explain query to get nullability: 

> After preparing, the Postgres driver will first look up the result columns in their source table and check if they have 
a `NOT NULL` constraint. Then, it will execute `EXPLAIN (VERBOSE, FORMAT JSON) <your query>` to determine which columns come from half-open joins (LEFT and RIGHT joins), which makes a normally `NOT NULL` column nullable. Since the `EXPLAIN VERBOSE` format is not stable or completely documented, this inference isn't perfect. However, it does err on the side of producing false-positives (marking a column nullable when it's `NOT NULL`) to avoid errors at runtime.

https://github.com/launchbadge/sqlx/blob/main/FAQ.md#how-do-the-query-macros-work-under-the-hood

**Comment by giacomocavalieri** [comment]
So I'm turning this into a discussion since I believe there's no way to tell if a query parameter is bound to a nullable column (not even by looking at the query explanation).
So this needs a bit more care and consideration to make its way into squirrel. I'm really interested in adding a way to tell squirrel if a query parameter should be nullable; some considerations:
- it cannot be a custom extension to the sql syntax (like `$1?`, ...) because it would go against squirrel's philosophy of "it's just plain SQL you can reuse everywhere"
- so it will probably need to be given in the form of some special comment before the query

Louis pointed out we could first ask the Postrgres folks if such an API to solve our problem could be added, that would for sure be the best possible outcome! I think I'll reach out soon and see how that goes

**Reply by binajmen** [reply]
> so it will probably need to be given in the form of some special comment before the query

I support the idea that one should be able to copy the exact query. Would a `-- nullable` be acceptable ?
```sql
insert into transactions (
    account,
    booking_date,
    statement_number,
) values (
    $1,
    $2,
    $3, -- nullable
)
returning id
```

**Reply by giacomocavalieri** [reply]
No I don't think it would, understanding the comment refers to the argument it precedes would require to parse the sql query which is a huge undertaking. O was thinking more of a comment preceding the query

**Reply by xxKeefer** [reply]
what about something like? 
```sql
-- nullable: $2, $3
insert into transactions (
    account,
    booking_date,
    statement_number,
) values (
    $1,
    $2,
    $3, 
)
returning id;
```

**Reply by binajmen** [reply]
In another discussion I suggested something in these lines:
```sql
-- type Transaction                <-- to rename the result from `TransactionRow` to `Transaction`
-- $1 account                      <-- to rename the function parameter from `arg_1` to `account`
-- $2 booking_date, nullable       <-- to rename + mark it nullable
-- $3 statement_number, nullable
```
Obviously, easier said than done 😅

**Comment by ross-byrne** [comment]
I'm currently switching my backend to use Squirrel (it's really awesome!) and I've run into this issue as well. My current work around is to add multiple queries where an input could be optional but this doesn't scale well.

> Louis pointed out we could first ask the Postrgres folks if such an API to solve our problem could be added, that would for sure be the best possible outcome! I think I'll reach out soon and see how that goes

Any updates on this?

**Reply by giacomocavalieri** [reply]
Nope, sorry! I haven't done that yet 🙇

**Comment by thorhj** [comment]
I too have run into this problem. For my case I can work around it with multiple queries, which is of course not ideal and perhaps unfeasible for larger projects. I also think you are right in being hesitant on the comments approach.

**Reply by giacomocavalieri** [reply]
Can you share the query? Maybe there's a different way to structure it to avoid the repetition

**Reply by thorhj** [reply]
My case was simply an insert query:

```
insert into poll (id, title, description)
values ($1, $2, $3)
```

In this case `description` is a nullable nvarchar, so the generated code would ideally use `Option(String)`. Instead I have to check that `description` has a value and use the above query, or use a similar query without `description` otherwise.

**Comment by DisguisedPigeon** [comment]
Hey, I've run into this as well writing a select query.

I'm here to show my temporal solution, and provide some more info for the discussion. I think the solution might be useful, since it avoids the exponentiallity of query duplication.

This is the query (simplified to 2 parameters):
```sql
select id from table
where ($1::int is null or key1 = $1)
  and ($2::text is null or key2 like $2)
  -- The type annotation (param::type) is needed because otherwise
  -- postgres isn't able to infer types due to the "is null" check.
 
  -- This might make the schema query solution not viable, since we
  -- no longer have assurance the schema's type nullability is kept
  -- during interpretation, and would need some kind of parser to
  -- make it work.  
```

I'm creating a new `custom_sql` module with the queries, using the squirrel generated types and editing the generated query manually to add nullability. This solves the exponential work issue, letting go of squirrel's type-safe queries in exchange.

Here are the relevant parts of the previous query's implementation. The comments detail the changes needed:
- generated code:
```gleam
// sql.gleam
pub type SampleQueryRow {
  SampleQueryRow(id: Int)
}

pub fn sample_query(
  db: pog.Connection,
  arg_1: Int,
  arg_2: String,
) -> Result(pog.Returned(SampleQueryRow), pog.QueryError) {
  // ...
}
```
- query function:
```gleam
// custom_sql.gleam

// Import the types and constructors from sql.gleam.
import entomologist/internal/sql.{
  type SampleQueryRow, SampleQueryRow
}

// Imports from sql.gleam.
import pog

// Custom function. Copied implementation from the generated one, but with some changes
pub fn sample_query(
  db: pog.Connection,
  // Add Option type to any nullable arguments
  arg_1: Option(Int),
  arg_2: Option(String),
) -> Result(pog.Returned(SampleQueryRow), pog.QueryError) {
  let decoder = {
    // You might need to add nullable to these fields. I'm not sure, maybe squirrel infers this correctly.
    use id <- decode.field(0, decode.int)
    decode.success(SampleQueryRow(id:))
  }

    "select id from some_table
where ($1::int is null or key1 = $1)
  and ($2::text is null or key2 like $2)
"
  |> pog.query
  // Add pog.nullable for the fields.
  |> pog.parameter(pog.nullable(pog.int, arg_1))
  |> pog.parameter(pog.nullable(pog.text, arg_2))
  |> pog.returning(decoder)
  |> pog.execute(db)
}
```

Hope this helps :)
